### ISSUE TITLE
Question on unique keys generation per operator instance

### ISSUE THREAD

**Issue by jczaja** [PaddlePaddle/Paddle#10912]
Hi,

We are working on concept of storing/reusing once created MKLDNN primitives. This challenge require to generate key for storing data in device context. Currently key is a hash value generated in a quite complex way . which will be difficult to maintain as operators are modified eg. generating key name based on params of operator and sizes of input.  To make things easier it would be good to have 
better method to generate keys that will be unique to the instance of operator. One considered way is 
to used name of output tensor as key eg. ctx.op().Output("Out") . However this is not good enough, as it fails to work for more complex models like machine_translation.py where we have RNNs and block operators. For example: ctx.op().Output("Out") evaluates to "fc.tmp.2" in softmax operator in machine_translation.py . But it is not unique as  we can inspect that there will be number of instances of softmax operator with the same name of output tensor but with diffrent input shape eg.  (16,30000 fc.tmp.2)  , (15, 30000  fc.tmp.2) (14, 30000 fc.tmp.2) etc.   Hence ctx.op().Output("Out") is not good enough.  Please advice on key name generation that is guaranteed to be unique within considered iteration .

**Comment by luotao1** [comment]
> But it is not unique as we can inspect that there will be number of instances of softmax operator with the same name of output tensor but with diffrent input shape eg. (16,30000 fc.tmp.2) , (15, 30000 fc.tmp.2) (14, 30000 fc.tmp.2) etc

@jczaja Could you explain more about this example? Do you mean
```
        fc
      /      \
softmax1   softmax2
```
And the key in softmax op or fc op is duplicated? In my mind, the name of output tensor of an op is unique.

**Comment by QiJune** [comment]
A unique name for every operator instance is necessary for naming inputs, logging, visualization and so on. I think it's really useful in many circumstances.

But if we add another required field in opdesc proto message, it will cause back compatibility problem. I am not sure if other pre-trained models will crash.

**Comment by jczaja** [comment]
@luotao1 To observe situation I'm describing please follow steps:
1) modify Paddle/benchmark/fluid/machine_translation.py to use mkldnn for softmax activation:
....
             out = fluid.layers.fc(input=h,
                                   size=target_dict_dim,
                                   bias_attr=True,
\+                                  **use_mkldnn=True,**
                                   act='softmax')
             rnn.output(out)
         return rnn()
....
2) Add in diagnostic info in softmax_mkldnn_op.cc:

   @@ -40,6 +40,9 @@ class SoftmaxMKLDNNKernel : public paddle::framework::OpKernel<T> {
     auto mkldnn_engine = dev_ctx.GetEngine();
     const Tensor* input = ctx.Input<Tensor>("X");
     Tensor* output = ctx.Output<Tensor>("Out");

\+    **std::cout << "Softmax MKLDNN Op  out_name: " << ctx.op().Output("Out") << " Dims(" << input->dims()[0] << ","<< in**
     PADDLE_ENFORCE(input->dims().size() == 2UL,
                    "The input of softmax op must be a 2D matrix.");
...

3) Rebuild PaddlePaddle. And run machine translation:

`python machine_translation.py --device CPU -iterations=1`

4) Expected output:
----------- seq2seq Configuration Arguments -----------
batch_size: 16
beam_size: 3
decoder_size: 512
device: CPU
dict_size: 30000
embedding_dim: 512
encoder_size: 512
infer_only: False
iterations: 1
learning_rate: 0.0002
max_length: 250
pass_num: 2
skip_batch_num: 5
with_test: False
------------------------------------------------
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(16,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(16,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(16,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(16,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(16,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(16,30000)
**Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(16,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(15,30000)**
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(13,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(13,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(12,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(12,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(12,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(12,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(10,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(9,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(8,30000)
Softmax MKLDNN Op  out_name: fc_10.tmp_2 Dims(8,30000)
...

As you can see (bold font) N dim does change from 16 to 15 , while output name remains the same eg. **fc_10.tmp_2**  

Hence, just taking Output("Out") as a key for caching MKLDNN primitives is not good enough.
So:
a) I would like to know why input->dims() do differ for same output tensor name. Is it because unrolling of Recurrent nets , or some scopes related problem? I'm not familiar how PaddlePadle is implementing RNNs or scopes so this is my guess.
b) Please advice on unique name for caching MKLDNN primitives.

**Comment by shanyi15** [comment]
您好，此issue在近一个月内暂无更新，我们将于今天内关闭。若在关闭后您仍需跟进提问，可重新开启此问题，我们将在24小时内回复您。因关闭带来的不便我们深表歉意，请您谅解~感谢您对PaddlePaddle的支持! 
 Hello, this issue has not been updated in the past month. We will close it today for the sake of other user‘s experience. If you still need to follow up on this question after closing, please feel free to reopen it. In that case, we will get back to you within 24 hours. We apologize for the inconvenience caused by the closure and thank you so much for your support of PaddlePaddle Group!
